### DISCUSSION TITLE
Custom exceptions throwing error

### DISCUSSION THREAD

**Discussion by ShocKwav3** [micropython/micropython#12693] [ESP8266] [answered]
Hi!

I have been stressing with this issue I am facing about my defined exception classes. As an example:
My custom defined class:

```
class SomeExceptionClass(Exception):
    _value = ''

    def __init__(self, value):
        self._value = value
        super().__init__()

    def __str__(self):
        return(repr(self._value))

```

I try to raise and handle it:

```
try:
    raise SomeExceptionClass('someMessage')
except SomeExceptionClass as e:
    print('oops')

```

I get the following error:

```
Traceback (most recent call last):
  File "boot.py", line 17, in <module>
TypeError: unsupported types for : 'TypeError', 'module'
```

I can not get my head around why is this issue happening. I tried searching but nothing fruitful came out. In fact, as I see similar post in the old micropython blog, it seemed it should work.  I tried in wokwi simulator and it worked fine. Although, the simulator was using esp32 with micropython 1.19.1.

My board is a esp01 and micropython version is 1.20.0

I'd appreciate any help on this matter!

**Comment by shariltumin** [comment]
You can try this:

```py
class myexception(Exception):
    def __init__(self, message):
        print('My Exception:')
        super().__init__('   '+message)
```

With that you can do:

```py
>>> 
MPY: soft reboot
MicroPython v1.22.0-preview.1.g523284777.dirty on 2023-10-07; ESP8266 module with ESP8266
>>> 
paste mode; Ctrl-C to cancel, Ctrl-D to finish
=== class myexception(Exception):
===     def __init__(self, message):
===         print('My Exception:')
===         super().__init__('   '+message)
=== 
=== try:
===    raise myexception('not ok that')
=== except Exception as e:
===    print(e)
=== 
My Exception:
   not ok that
>>>  
```

**Reply by ShocKwav3** [reply]
In your code, `myexception` was raised but you have `except Exception as e:` to catch it. I am trying to catch a specific exception such as `myexception` in your case like `except myexception as e:`. Is that not possible in micropython?

**Reply by shariltumin** [reply]
```
>>> try:
...     raise myexception('not ok that')
... except myexception as e:
...     print(e)
... 
My Exception:
   not ok that
```
```

**Reply by ShocKwav3** [reply]
Didn't work.

```
class myexception(Exception):
    def __init__(self, value):
        super().__init__('   '+value)
```
```
try:
    raise myexception('errorMessage')
except myexception as e:
    print('oops')
```
I get:

```
MPY: soft reboot
Traceback (most recent call last):
  File "boot.py", line 19, in <module>
TypeError: unsupported types for : 'TypeError', 'module'
MicroPython v1.20.0 on 2023-04-26; ESP module (1M) with ESP8266
Type "help()" for more information.
```

**Reply by shariltumin** [reply]
What do you have on line 19 of the boot.py file?

**Reply by peterhinch** [reply]
It's OK here:
```python
Python 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> class myexception(Exception):
...     def __init__(self, value):
...         super().__init__('   '+value)
... 
>>> try:
...     raise myexception('errorMessage')
... except myexception as e:
...     print('oops')
... 
oops
>>> 
```

**Reply by ShocKwav3** [reply]
> What do you have on line 19 of the boot.py file?

@shariltumin bunch of commented code, I had it in project of mine. I did it in a single file now (please see my screenshot below) and still the same issue. 

@peterhinch not sure what is going on... I have tried pasting the code in the REPL and it works, as I try to run it in the device I mean putting it in the main.py or boot.py I get the error. Please see my screenshot.

<img width="500" alt="Screenshot 2023-10-16 at 19 24 36" src="https://github.com/micropython/micropython/assets/23660537/745e5058-3670-465a-bb37-c18fabff2264">

**Reply by shariltumin** [reply]
Can you try deleting line 8 (blank line)? There may be unprintable characters there.

And while you are in the REPL, try ```import main```. Do you get the same error?

**Reply by ShocKwav3** [reply]
@shariltumin found even more strange thing. I kept no line break and it works.

<img width="491" alt="Screenshot 2023-10-17 at 1 15 08" src="https://github.com/micropython/micropython/assets/23660537/5fa25833-9bbb-47a6-96d1-f6ddd29d4e57">

Then I put the exception in a different location and try to run the code and it fails.

<img width="471" alt="Screenshot 2023-10-17 at 1 12 49" src="https://github.com/micropython/micropython/assets/23660537/cfde9fb8-560a-42c1-a1b7-e1eb98cb6c10">


As seen in the screenshots, the `main.py` is in the root and the `wifiexception` is under `exceptions` folder.

What am I doing wrong here?

**Comment by jimmo** [comment, marked as answer]
@ShocKwav3 there's something about the way you're importing the exception from your `exceptions` directory...  Back in the original post, the error is a bit subtle, but tells you that instead of importing `SomeExceptionClass`, you imported a module, but named it `SomeExceptionClass`. Perhaps you wrote

```
import something as SomeExceptionClass
```
??

Or perhaps something in `exceptions/__init__.py` is doing something strange?

The error you listed was:

```
Traceback (most recent call last):
  File "boot.py", line 17, in <module>
TypeError: unsupported types for : 'TypeError', 'module'
```

This means that it is trying to find an except block for an active `TypeError`, and it's trying to match it against something (`SomeExceptionClass`) which is of type `module`. The `TypeError` comes from trying to `raise` something that isn't an exception.

This is why it works when you run it directly (as what you're donig is definitely supported as @shariltumin and @peterhinch have pointed out), but not when you import the exception.

**Reply by ShocKwav3** [reply]
@jimmo Yes, you are right. I did have `import something as SomeExceptionClass`.

> The TypeError comes from trying to raise something that isn't an exception.

I see what happened now... Thanks for the explanation and pointing out the mistake.

**Comment by shariltumin** [comment]
As Jim pointed out earlier, your import statement may be in error.

Since you have a subdirectory called "exceptions" and you put the wifiexception.py file there, please try.


```py
from exceptions.wifiexception import wifiexception
```

The above tries to import wifiexception from "./exceptions/wifiexception.py".

I will do it differently. I will have a file "my_exceptions.py". I will declare all my all my exception classes in it. My import will be


```py
from my_exceptions import myexception1, myexception2
```

It can be confusing to name a module file and a class in the modules with the same name.

**Reply by ShocKwav3** [reply]
@shariltumin Yes, the issue turned out to be the way I was importing it. Now I got it working. Thank you for taking time to debug this with me. Cheers!
